skactiveml/pool/_badge.py

```diff
@@ -0,0 +1,232 @@
+        check_scalar(fit_clf, "fit_clf", bool)
+        check_scalar(return_embeddings, "return_embeddings", bool)
+
+        # Fit the classifier
+        if fit_clf:
+            clf = clone(clf).fit(X, y)
+
+        # find the unlabeled dataset
+        if candidates is None:
+            X_unlbld = X_cand
+            unlbld_mapping = mapping
+        elif mapping is not None:
+            unlbld_mapping = unlabeled_indices(
+                y[mapping], missing_label=self.missing_label
+            )
+            X_unlbld = X_cand[unlbld_mapping]
+            unlbld_mapping = mapping[unlbld_mapping]
+        else:
+            X_unlbld = X_cand
+            unlbld_mapping = np.arange(len(X_cand))
+
+        # gradient embedding, aka predict class membership probabilities
+        if return_embeddings:
+            probas, X_unlbld = clf.predict_proba(
+                X_unlbld, return_embeddings=True
+            )
+        else:
+            probas = clf.predict_proba(X_unlbld)
+        p_max = np.max(probas, axis=1).reshape(-1, 1)
+        g_x = (p_max - 1) * X_unlbld
+
+        # init the utilities
+        if mapping is not None:
+            utilities = np.full(
+                shape=(batch_size, X.shape[0]), fill_value=np.nan
+            )
+        else:
+            utilities = np.full(
+                shape=(batch_size, X_cand.shape[0]), fill_value=np.nan
+            )
+
+        # sampling with kmeans++
+        query_indicies = np.array([], dtype=int)
+        query_indicies_in_unlbld = np.array([], dtype=int)
+        d_2_s = []
+        for i in range(batch_size):
+            if i == 0:
+                d_2 = _d_2(g_x, [])
+            else:
+                d_2 = _d_2(g_x, [idx_in_unlbld], d_2_s[i - 1])
+            d_2_s.append(d_2)
+
+            d_2_sum = np.sum(d_2)
+            if i == 0 or d_2_sum == 0:
+                if d_2_sum == 0:
+                    d_2_s[-1] = np.full(shape=len(g_x), fill_value=np.inf)
+                d_2 = np.ones(shape=len(g_x))
+                d_2[query_indicies_in_unlbld] = 0
+                d_2_sum = np.sum(d_2)
+
+            d_probas = d_2 / d_2_sum
+
+            utilities[i, unlbld_mapping] = d_probas
+            utilities[i, query_indicies] = np.nan
+
+            idx_in_unlbld_array = self.random_state_.choice(
+                len(d_probas), 1, replace=False, p=d_probas
+            )
+            idx_in_unlbld = idx_in_unlbld_array[0]
+            query_indicies_in_unlbld = np.append(
+                query_indicies_in_unlbld, idx_in_unlbld_array
+            )
+
+            idx = unlbld_mapping[idx_in_unlbld]
+            query_indicies = np.append(query_indicies, [idx])
+
+        if return_utilities:
+            return query_indicies, utilities
+        else:
+            return query_indicies
+
+
+def _d_2(g_x, query_indicies, d_latest=None):
+    """
+    Calculates the D^2 value of the embedding features of unlabeled
+    data.
+
+    Parameters
```

Comment:
Check spelling and dots at the end of parameter descriptions. Further, there should be space between the parameter name and the :